packages/components/src/number-control/index.tsx

```diff
@@ -180,33 +194,60 @@ function UnforwardedNumberControl(
 			 * Handles commit (ENTER key press or blur)
 			 */
 			if (
-				type === inputControlActionTypes.PRESS_ENTER ||
-				type === inputControlActionTypes.COMMIT
+				action.type === inputControlActionTypes.PRESS_ENTER ||
+				action.type === inputControlActionTypes.COMMIT
 			) {
 				const applyEmptyValue =
-					required === false && currentValue === '';
+					currentValue === undefined ||
+					( required === false && currentValue === '' );
 
-				// @ts-expect-error TODO: Resolve discrepancy between `value` types in InputControl based components
 				nextState.value = applyEmptyValue
-					? currentValue
-					: // @ts-expect-error TODO: Investigate if it's ok for currentValue to be undefined
```

Comment:
Solved this TODO by adding the `currentValue === undefined` check when computing the `applyEmptyValue` variable a few lines above